fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1129,6 +1122,23 @@ def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
 
         return outputs
 
+    def _checkpoint_module_last_backward_call(self) -> bool:
+        # If this is a checkpointed FSDP module, e.g. checkpoint(FSDP(module)),
+        # we check if the following counter reaches 0. If it is, it is the last
+        # inner backward call for this FSDP module.
+        return getattr(self._fsdp_wrapped_module, "_checkpoint_fwd_counter", 0) == 0
+
+    def _require_final_backward(self) -> bool:
+        assert self._is_root
+        for m in self.modules():  # includes self
```

Comment:
I am not sure the purpose of this loop. If you need to check whether any trainable parameters exists inside the root, just m.parameters() would be good enough? No need to loop over all instances?

Reply:
that is true, we do not need to loop here!